Add support for persistent subscription negative acknowledgments.

nathanfox opened the pull request:
In my particular use case I require being able to send negative acknowledgments for persistent subscription event messages.  This code adds handling for sending PersistentSubscriptionNakEvents messages.

Some notes about the code.

I added a unit test which tests sending a Retry nak. The test is *Try retry nak action* in file [extreme_test.exs](https://github.com/nathanfox/extreme/blob/f3e5b2ccba83d53d3f2b5fa3dbb2a215d33d991f/test/extreme_test.exs) line 720.

I added the ability to delete a persistent subscription. This allows for cleanup of created subscriptions in the test code.

In response.ex, the message SubscriptionDropped was commented out, but I wanted to handle this message for a persistent subscription, so I added it back. I'm not sure if this is a problem otherwise? All unit test pass and I haven't had problems in my system testing.

@slashdotdash could you please take a look at the code and let me know what you think? I really appreciate it.

If any changes are needed or I can answer any questions, please let me know.

Thank you,
Nathan Fox

Commit:
Add support for persistent subscription negative acknowledgments.

nathanfox commented:
The CI run failed for the merge request, but the tests run fine on my machine. I'm not sure why it failed :-(

slashdotdash commented:
The Travis CI test failure appears to be unrelated to your changes (listener test failure, not persistent subscriptions). 

@nathanfox I would recommend adding the following `nak/5` function to support persistent subscriptions on projection streams. You have to ack/nak the linked event in this scenario.

```elixir
def nak(subscription, %{link: link}, correlation_id, nak_action, message \\ nil) when not is_nil(link) and is_atom(nak_action) do
  GenServer.call(subscription, {:nak, link.event_id, correlation_id, nak_action, message})
end
```

Otherwise looks good to me, thanks for adding naks.

burmajam commented:
I would only suggest to rename **nak** to **nack**, but overall good job!

nathanfox commented:
Good point on the name change. I'll change that. I'll also work on @slashdotdash 's suggestion. I'm not sure if I can get to it this week. Hopefully by early next week I'll have the changes in place and I'll update the branch in the pull request.

burmajam commented:
Great, take your time :)

Commit:
Rename nak to nack.

Commit:
Add nack function to support persistent subscriptions on projected streams.
- Handle nack for linked event.
- Add unit test for retry of linked events.

nathanfox commented:
I've added the nak to nack name change as well as adding a nack function to handle linked events.

Please let me know how it looks.

slashdotdash commented:
@nathanfox The example in the README is using `nak`, not the renamed `nack` function.

Commit:
Update wording in readme for not acknowledged events.
- The correct terminology is not acknowledged.
- Updated the text and sample code.

nathanfox commented:
Good catch. I ran grep against .ex and .exs :-) I updated the readme. I also corrected the terminology so nack means not acknowledged to match the EventStore documentation.
